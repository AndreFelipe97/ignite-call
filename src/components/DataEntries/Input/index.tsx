import { TextInput } from "@ignite-ui/react";

interface InputProps {
  size?: "sm" | "md";
  prefix?: string;
  placeholder?: string;
  name: string;
  register: any;
}

export function Input({
  size = "sm",
  prefix = "",
  placeholder = "",
  name,
  register,
}: InputProps) {
  return (
    <TextInput
      size={size}
      prefix={prefix}
      placeholder={placeholder}
      {...(register && register(name))}
    />
  );
}
