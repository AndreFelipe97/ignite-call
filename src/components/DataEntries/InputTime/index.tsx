import { TextInput } from "@ignite-ui/react";

interface InputProps {
  name: string;
  isDisabled: boolean;
  register: any;
}

export function InputTime({ name, isDisabled, register }: InputProps) {
  return (
    <TextInput
      size="sm"
      type="time"
      step={60}
      disabled={isDisabled}
      data-testid={name}
      {...(register && register(name))}
    />
  );
}
