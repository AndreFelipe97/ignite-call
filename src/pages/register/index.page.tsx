import { Heading, MultiStep, Text } from "@ignite-ui/react";
import { useForm } from "react-hook-form";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";
import { ArrowRight } from "phosphor-react";
import { Container, Form, FormError, Header } from "./styles";
import { useRouter } from "next/router";
import { useEffect } from "react";
import { api } from "@/lib/axios";
import { AxiosError } from "axios";
import { SubmitButton } from "@/components/Buttons/SubmitButton";
import { Input } from "@/components/DataEntries/Input";

const registerFormSchema = z.object({
  username: z
    .string()
    .min(3, { message: "O usuário precisar ter três letras." })
    .regex(/^([a-z\\-]+)$/i, {
      message: "O usuário pode conter apenas letras e hifens.",
    })
    .transform((username) => username.toLowerCase()),
  name: z
    .string()
    .min(3, { message: "O nome precisar ter três letras." })
    .regex(/^([a-z]+)$/i, {
      message: "O nome pode conter apenas letras.",
    }),
});

type registerFormData = z.infer<typeof registerFormSchema>;

export default function Register() {
  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors, isSubmitting },
  } = useForm<registerFormData>({
    resolver: zodResolver(registerFormSchema),
  });

  const router = useRouter();

  useEffect(() => {
    if (router.query.username) {
      setValue("username", String(router.query.username));
    }
  }, [router.query?.username, setValue]);

  async function handleRegister(data: registerFormData) {
    try {
      await api.post("/users", {
        name: data.name,
        username: data.username,
      });

      await router.push("/register/connect-calendar");
    } catch (err) {
      if (err instanceof AxiosError && err?.response?.data?.message) {
        alert(err.response.data.message);
        return;
      }
      console.error(err);
    }
  }

  return (
    <Container>
      <Header>
        <Heading as="strong">Bem-vindo ao ignite call!</Heading>
        <Text>
          Precisamos de algumas informações para criar seu perfil! Ah, você pode
          editar essas informações depois
        </Text>
        <MultiStep size={4} currentStep={1} />
      </Header>
      <Form as="form" onSubmit={handleSubmit(handleRegister)}>
        <label>
          <Text size="sm">Nome de usuário</Text>
          <Input
            size="sm"
            prefix="ignite.com/"
            placeholder="seu-usário"
            name="username"
            register={register}
          />
          {errors.username && (
            <FormError size="sm">{errors.username.message}</FormError>
          )}
        </label>
        <label>
          <Text size="sm">Nome completo</Text>
          <Input
            size="sm"
            placeholder="Seu nome"
            name="name"
            register={register}
          />
          {errors.name && (
            <FormError size="sm">{errors.name.message}</FormError>
          )}
        </label>
        <SubmitButton
          label="Próximo passo"
          isDisabled={isSubmitting}
          icon={<ArrowRight />}
        />
      </Form>
    </Container>
  );
}
