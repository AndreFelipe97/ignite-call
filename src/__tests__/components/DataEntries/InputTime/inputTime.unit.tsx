import { render, screen } from "@testing-library/react";
import { InputTime } from "@/components/DataEntries/InputTime";
import { register } from "@/__factories__/register";

describe("<InputTime />", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly when is enabled", () => {
    render(
      <InputTime
        isDisabled={false}
        name="test-input-time"
        register={register}
      />
    );

    const inputTime = screen.getByTestId("test-input-time");

    expect(inputTime).toBeInTheDocument();
    expect(register).toBeCalledTimes(1);
    expect(inputTime).toBeEnabled();
    expect(inputTime).toMatchSnapshot();
  });

  it("should render correctly when is disabled", () => {
    render(
      <InputTime isDisabled={true} name="test-input-time" register={register} />
    );

    const inputTime = screen.getByTestId("test-input-time");

    expect(inputTime).toBeInTheDocument();
    expect(register).toBeCalledTimes(1);
    expect(inputTime).toBeDisabled();
    expect(inputTime).toMatchSnapshot();
  });
});
