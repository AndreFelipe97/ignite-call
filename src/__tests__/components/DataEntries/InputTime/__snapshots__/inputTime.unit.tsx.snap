// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputTime /> should render correctly when is disabled 1`] = `
<input
  class="c-grKefI"
  data-testid="test-input-time"
  disabled=""
  step="60"
  type="time"
/>
`;

exports[`<InputTime /> should render correctly when is enabled 1`] = `
<input
  class="c-grKefI"
  data-testid="test-input-time"
  step="60"
  type="time"
/>
`;
